src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -0,0 +1,186 @@
+  }
+
+  public SubmitExerciseAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                              @NotNull Notifier notifier) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.notifier = notifier;
+  }
+
+  @Override
+  public void update(@NotNull AnActionEvent e) {
+    MainViewModel mainViewModel = mainViewModelProvider.getMainViewModel(e.getProject());
+    CourseViewModel courseViewModel = mainViewModel.courseViewModel.get();
+    ExercisesTreeViewModel exercisesTreeViewModel
+        = mainViewModel.exercisesViewModel.get();
+    APlusAuthenticationViewModel authenticationViewModel
+        = mainViewModel.authenticationViewModel.get();
+
+    e.getPresentation().setEnabled(e.getProject() != null && exercisesTreeViewModel != null
+        && authenticationViewModel != null && courseViewModel != null
+        && courseViewModel.getModel() != null);
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    MainViewModel mainViewModel = mainViewModelProvider.getMainViewModel(e.getProject());
+    ExerciseViewModel selectedExercise
+        = mainViewModel.exercisesViewModel.get().getSelectedExercise();
+    if (selectedExercise == null) {
+      return;
+    }
+
+    APlusAuthentication authentication
+        = mainViewModel.authenticationViewModel.get().getAuthentication();
+    Course course = mainViewModel.courseViewModel.get().getModel();
+    Project project = e.getProject();
+
+    SubmittableExercise exercise = tryGetSubmittableExercise(
+        selectedExercise.getModel(), authentication, project);
+    if (exercise == null) {
+      return;
+    }
+
+    SubmissionHistory submissionHistory = tryGetSubmissionHistory(
+        exercise, authentication, project);
+    if (submissionHistory == null) {
+      return;
+    }
+
+    List<Group> groups = tryGetUserGroups(course, authentication, project);
+    if (groups == null) {
+      return;
+    }
+
+    SubmissionViewModel viewModel = new SubmissionViewModel(
+        exercise, submissionHistory, groups, authentication, project);
+
+    if (!new ModuleSelectionDialog(viewModel).showAndGet()
+        || viewModel.getSelectedModule() == null) {
+      return;
+    }
+
+    List<Path> filePaths = tryGetFilePaths(
+        exercise.getFilenames(), viewModel.getSelectedModule(), project);
+    if (filePaths == null) {
+      return;
+    }
+
+    if (new SubmissionDialog(viewModel).showAndGet()) {
+      // Do actual submission
+    }
+  }
+
+  @Nullable
+  private SubmittableExercise tryGetSubmittableExercise(@NotNull Exercise exercise,
+                                                        @NotNull APlusAuthentication authentication,
+                                                        @Nullable Project project) {
+    try {
+      SubmittableExercise submittableExercise = SubmittableExercise.fromExerciseId(
+          exercise.getId(), authentication);
```

Comment:
Cannot we somehow reuse `exercise` object we already have?

Reply:
Do you mean specifically in the action here, or do you mean the class `SubmittableExercise` in general?

`SubmittableExercise` includes the names of the files to be submitted which aren't available in `Exercise`.